### PR TITLE
1051: Add local invoke to proxies that trace and record LB

### LINKED ISSUE #1051: Add local invoke to proxies that trace and record LB
**What Needs to be Done?**

For objgroups, collections, (and maybe bare handlers), add a local proxy invoke. Right now, for instance, you can do `proxy.send<MyCol, &MyCol::handler>()` which always goes through the scheduler even if it's local. So if you want to execute right away, you have run the handler manually. Add a `proxy.invoke<...>()` that does the local invocation inline. It should error if the element does not exist where the user tried to invoke it.

### PR DESCRIPTION
**Pull request by JacobDomagala**
Fixes #1051 

- [x] virtual collection proxy
- [x] obj group proxy
- [x] unittests

### CODE REVIEW

File: src/vt/trace/trace_constants.h [hunk 1]
```diff
@@ -83,7 +83,8 @@ enum class eTraceConstants : int32_t {
   UserStat                  = 32,
   BeginUserEventPair        = 98,
   EndUserEventPair          = 99,
-  UserEventPair             = 100
+  UserEventPair             = 100,
+  LocalInvoke               = 101
```
Comment (JacobDomagala) [inline src/vt/trace/trace_constants.h:87]: Not sure whether we want (or can as this probably won't work with Projections tool) to add new trace type for invoke or should we use existing one? At the end the output will be the same as for Creation event type.

**Comment by lifflander** [comment]
Per @PhilMiller suggestion, we should create a `makeTraceCreationSend(..)` for both starting the local invocation and on the receiving side after the invocation returns. So it will looks like a message was sent to start the local invoke and then a message was sent back to the same node to continue the work after the local invocation.

File: src/vt/trace/trace_constants.h (hunk 1 above)
Comment (lifflander) [reply src/vt/trace/trace_constants.h:87]: Let's make this just a regular `Creation` event.

Comment (lifflander) [reply src/vt/trace/trace_constants.h:87]: Projections doesn't do anything special for these.

File: src/vt/vrt/collection/invoke/invokable.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+#if !defined INCLUDED_VRT_COLLECTION_INVOKE_INVOKABLE_H
+#define INCLUDED_VRT_COLLECTION_INVOKE_INVOKABLE_H
+
+#include "vt/config.h"
+#include "vt/vrt/collection/active/active_funcs.h"
+
+namespace vt { namespace vrt { namespace collection {
+
+template <typename ColT, typename IndexT, typename BaseProxyT>
```
Comment (lifflander) [inline src/vt/vrt/collection/invoke/invokable.h:53]: Can you add doxygen for these new methods here?

**Review by lifflander — APPROVED** [review]
Overall, this looks great and it's nearly ready to merge I think. Just a little more doxygen to be added to the proxy interface for virtual collections. Also, we should just use the standard trace event.

**Comment by JacobDomagala** [comment]
@PhilMiller @lifflander  Currently invoke produces following traces:

> 1. Creation event (let's say with trace_id == 2)
> 2. Begin Processing (trace_id == 2)
> _execute user's function_
> 3. End Processing (trace_id == 2)


So you're proposing we should add another Creation event like following?

> 1. Creation event (let's say with trace_id == 2)
> 2. Begin Processing (trace_id == 2)
> _execute user's function_
> 3. End Processing (trace_id == 2)
> 4. Creation event (**trace_id == 3**)

Or do you have something else in mind?

**Comment by JacobDomagala** [comment]
PR updated

**Comment by lifflander** [comment]
> PR updated

Looks good I think this is ready to rebase and merge

**Pull request merged**